### PR TITLE
Fix group shares on v1 caldav and carddav

### LINKED ISSUE #23273: Shared addressbook is always writable independently on "can edit" checkbox
### Steps to reproduce
1. Share addressbook with group (I used LDAP authentication)
2. Remove "can edit" check
3. Login with user that should have read-only access and drop contacts from addressbook

### Expected behaviour
AB obviously should be read only

### Actual behaviour
Any user can drop contents of shared addressbook

**Database:** mariadb

**PHP version:** 7.0.4

**ownCloud version:** (see ownCloud admin page) 9.0.0

**Updated from an older ownCloud or fresh install:** updated many times from some ancient version

Also I tried to change access to 2 or 3 in oc_dav_shares, but AB editable anyway

### LINKED ISSUE #23306: current-user-principal returns incorrect href.
Request the current user principal from carddav and it returns an incorrect href for the current user principal on owncloud 9:

```
$ curl  -u myself --data @propfind4.xml -H 'Depth: 0' -H 'Content-Type: application/xml; charset-utf-8' -X PROPFIND https://www.example.com/owncloud/remote.php/carddav/
```
with propfind4.xml:
```
<d:propfind xmlns:d="DAV:">
  <d:prop>
     <d:current-user-principal />
  </d:prop>
</d:propfind>
```
and after entering correct user password, returns:
```
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:card="urn:ietf:params:xml:ns:carddav">
 <d:response>
  <d:href>/owncloud/remote.php/carddav/</d:href>
  <d:propstat>
   <d:prop>
    <d:current-user-principal>
     <d:href>/owncloud/remote.php/carddav/principals/users/myself/</d:href>
    </d:current-user-principal>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
</d:multistatus>
```
However, /owncloud/remote.php/carddav/principals/users/myself/ does not exist. The only principal uri that exists would be /owncloud/remote.php/carddav/principals/myself/

### LINKED ISSUE #23328: shared address book doesn't show through carddav
owncloud 9: If someone shares an address book with me, I can see the contacts from the address book online in the contacts app. However, through carddav the address book is not shown. If I do a:
```
$ curl -u myself --data @propfind3.xml -H 'Depth: 1' -H 'Content-Type: application/xml; charset-utf-8' -X PROPFIND https://www.example.com/owncloud/remote.php/carddav/addressbooks/myself/
```
with propfind3.xml containing:
```
<d:propfind xmlns:d="DAV:" xmlns:cs="http://calendarserver.org/ns/">
  <d:allprop/>
</d:propfind>
```
I only get this:
```
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:card="urn:ietf:params:xml:ns:carddav">
 <d:response>
  <d:href>/owncloud/remote.php/carddav/addressbooks/myself/</d:href>
  <d:propstat>
   <d:prop>
    <d:resourcetype>
     <d:collection/>
    </d:resourcetype>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:href>/owncloud/remote.php/carddav/addressbooks/myself/contacts/</d:href>
  <d:propstat>
   <d:prop>
    <d:resourcetype>
     <d:collection/>
     <card:addressbook/>
    </d:resourcetype>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
</d:multistatus>
```
which is my own collection and my own address book "contacts" but the shared one is missing.

The problem seems to be that getAddressBooksForUser is called with principal "principals/myself" from the getChildren in UserAddressBooks which only returns my own address books.

(Another call to getAddressBooksForUser from the hookmanager uses principal "principals/users/mysql" which does in fact return my own and the shared address book. So it seems an issue of the principal uri used.)

### PR DESCRIPTION
**Pull request by DeepDiver1975**
@gvde please test - THX
# Test Scenario 1 - fixes https://github.com/owncloud/core/issues/23273
1. Share a calendar with another user readonly
2. See the calendar appear in the calendar app of the user you shared with
3. Try to edit -> should not be possible
# Test Scenario 2 - fixes https://github.com/owncloud/core/issues/23306
1. run propfind against the old carddav url

```
$ curl  -u myself --data @propfind4.xml -H 'Depth: 0' -H 'Content-Type: application/xml; charset-utf-8' -X PROPFIND https://www.example.com/owncloud/remote.php/carddav/
```

with propfind4.xml:

```
<d:propfind xmlns:d="DAV:">
  <d:prop>
     <d:current-user-principal />
  </d:prop>
</d:propfind>
```
1. the returned principal shall not contain '/users/' in between
# Test Scenario 3 - fixes https://github.com/owncloud/core/issues/23328
1. share a calendar and/or and addressbook with a user
2. the shared resource has to be listed in the old endpoints - use the dav browser plugin to verify

### CODE REVIEW

File: apps/dav/lib/connector/sabre/principal.php [hunk 1]
```diff
@@ -135,7 +135,7 @@ public function getGroupMemberSet($principal) {
 	 * @return array
 	 * @throws Exception
 	 */
-	public function getGroupMembership($principal) {
+	public function getGroupMembership($principal, $needGroups = false) {
```
Comment (LukasReschke) [inline apps/dav/lib/connector/sabre/principal.php]: PHPDoc

**Comment by gvde** [comment]
b47d25f11e4fbee17ea9b07bd69f48c636d6e8e2 now returns /principals/myself which exists.

The address book collection still only returns my own address book not the shared one...

**Comment by DeepDiver1975** [comment]
> The address book collection still only returns my own address book not the shared one...



> The address book collection still only returns my own address book not the shared one...

mind retesting? I assume you did checkout this branch and not only apply the commits on top of your oc9 installation

````
deepdiver@alien:~/Development/ownCloud/master [fix-group-sharing-for-v1-caldav-and-carddav]$ curl -u user01 --data @propfind3.xml -H 'Depth: 1' -H 'Content-Type: application/xml; charset-utf-8' -X PROPFIND http://localhost:8888/remote.php/carddav/addressbooks/user01/
Enter host password for user 'user01':
<?xml version="1.0"?>
<d:multistatus xmlns:d="DAV:" xmlns:s="http://sabredav.org/ns" xmlns:card="urn:ietf:params:xml:ns:carddav">
 <d:response>
  <d:href>/remote.php/carddav/addressbooks/user01/</d:href>
  <d:propstat>
   <d:prop>
    <d:resourcetype>
     <d:collection/>
    </d:resourcetype>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:href>/remote.php/carddav/addressbooks/user01/default_shared_by_admin/</d:href>
  <d:propstat>
   <d:prop>
    <d:resourcetype>
     <d:collection/>
     <card:addressbook/>
    </d:resourcetype>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
 <d:response>
  <d:href>/remote.php/carddav/addressbooks/user01/testAddr_shared_by_admin/</d:href>
  <d:propstat>
   <d:prop>
    <d:resourcetype>
     <d:collection/>
     <card:addressbook/>
    </d:resourcetype>
   </d:prop>
   <d:status>HTTP/1.1 200 OK</d:status>
  </d:propstat>
 </d:response>
</d:multistatus>

**Comment by gvde** [comment]
I don't have a test server thus I cannot really test branches but apply the patches instead...

Is there an issue for switching from these "shadow clones" of shared address books to using ACLs and user configurable address book collections (i.e. the user can decide which accessible address books to use and which not)? The current way of sharing doesn't really work well and doesn't scale when you share a lot of address books across users...

**Comment by DeepDiver1975** [comment]
> Is there an issue for switching from these "shadow clones" of shared address books to using ACLs and user configurable address book collections (i.e. the user can decide which accessible address books to use and which not)? The current way of sharing doesn't really work well and doesn't scale when you share a lot of address books across users...

let's discuss this in an issue on its own or better on irc - THX

**Comment by gvde** [comment]
> let's discuss this in an issue on its own or better on irc - THX

#23375

**Comment by DeepDiver1975** [comment]
caldav and carddav tests are back green again \o/

**Comment by LukasReschke** [comment]
:+1: once https://github.com/owncloud/core/pull/23342/files#r56494820 is fixed

File: apps/dav/lib/connector/sabre/principal.php (hunk 1 above)
Comment (DeepDiver1975) [inline apps/dav/lib/connector/sabre/principal.php]: addressed

**Comment by PVince81** [comment]
Code looks fine :+1: 

@gvde did you get a chance to test ?

**Comment by DeepDiver1975** [comment]
backport at https://github.com/owncloud/core/pull/23552

**Comment by PVince81** [comment]
@karlitschek backport to stable9 ?

**Comment by karlitschek** [comment]
makes sense. please backport :+1:

**Comment by gvde** [comment]
As I don't have a separate test server, I have applied the backport #23552 on my server running 9.0.1~beta-1.1 and tested it.

In regard to #23306 it is OK.

However, nothing has changed for me in regard to #23328. A PROPFIND for allprop on /owncloud/remote.php/carddav/addressbooks/username/ still only gets me the user address book but not the address book which has been shared to a group in which username is member. I still get the same output as posted in #23328...

**Comment by DeepDiver1975** [comment]
> However, nothing has changed for me in regard to #23328. A PROPFIND for allprop on /owncloud/remote.php/carddav/addressbooks/username/ still only gets me the user address book but not the address book which has been shared to a group in which username is member. I still get the same output as posted in #23328...

I honestly think there went something wrong with the way you applied the patches - but I happily retest this ...

**Comment by DeepDiver1975** [comment]
bildschirmfoto von 2016-03-24 16-02-51

bildschirmfoto von 2016-03-24 16-03-10

**Comment by DeepDiver1975** [comment]
the shared addressbook can even be accessed by the original url :wink: 

bildschirmfoto von 2016-03-24 16-05-26

**Comment by gvde** [comment]
@DeepDiver1975 
> have a look below - this is current master - I have admin's default addressbook shared with the user01 (and with the group as well - but this has no impact) and using cadaver the shared addressbook is listed

"no impact" in the sense of "it doesn't work" or of "it works the same".

I have just tried it: if the address book is shared with the user it works. If the address book is shared with a group in which the user is a member it doesn't work.

Thus, remove the share with 'user01' in your example above and check if it is still listed, if user01 belongs to group 'users'...

I have just retested it again: reinstalled 9.0.1~beta-1.1 (ubuntu 14.04), reapplied the patch for 23552. All patch (except the tests if course) were applied successfully, no reverses. So I would say the patch is correctly applied...

**Comment by DeepDiver1975** [comment]
> "no impact" in the sense of "it doesn't work" or of "it works the same".

we are talking about group sharing in this pr - so I better test this explicitly - right?

**Comment by gvde** [comment]
Hmm. What does cadaver to for "ls" on /carddav/ or /carddav/addressbooks/ ?

I only get a "Listing members of this collection is disabled" when I do a PROPFIND for allprop on that URIs...

**Comment by DeepDiver1975** [comment]
> I only get a "Listing members of this collection is disabled" when I do a PROPFIND for allprop on that URIs...

listing principals is disabled by default - set debug to true in config.php and it will be allowed

**Comment by MorrisJobke** [comment]
Also fixes #23011

**Pull request merged**